### PR TITLE
feat: adding gocover to e2e tests (#18130)

### LINKED ISSUE #18130: Collect test coverage from e2e tests
# Summary

Today we only collect coverage from unit tests. We should also collect coverage from e2e tests.

# Motivation

It would be helpful to get a clearer idea of our complete coverage.

# Proposal

New go versions support collecting coverage from running code. We could collect profiles from each of the components and merge them all with the unit coverage data. https://tip.golang.org/doc/go1.20#cover

### PR DESCRIPTION
**Pull request by rumstead**
closes #18130

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: .github/workflows/ci-build.yaml [hunk 1]
```diff
@@ -367,10 +367,13 @@ jobs:
         with:
           name: test-results
           path: test-results
+      - name: combine-go-coverage
+        run: |
+          go tool covdata percent -i=./test-results -o test-results/full-coverage.out
```
Comment (rumstead) [inline .github/workflows/ci-build.yaml:372]: Sorta closing my eyes and hoping those files made it and are in the test-results dir :). I see we downloaded the test-results dir but I don't see us doing anything with it. We get the `coverage.out` right after and upload that.

**Comment by crenshaw-dev** [comment]
Thanks for the PR! Can you rebase? Codecov report uploading was broken.

**Comment by crenshaw-dev** [comment]
Looks like codecov is uploading, but the coverage amount hasn't changed... will look tomorrow

File: Procfile [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
+controller: [ "$BIN_MODE" = 'true' ] && COMMAND=./dist/argocd || COMMAND='go run ./cmd/main.go' && sh -c "GOCOVERDIR="test-results" HOSTNAME=testappcontroller-1  FORCE_LOG_COLORS=1 ARGOCD_FAKE_IN_CLUSTER=true ARGOCD_TLS_DATA_PATH=${ARGOCD_TLS_DATA_PATH:-/tmp/argocd-local/tls} ARGOCD_SSH_DATA_PATH=${ARGOCD_SSH_DATA_PATH:-/tmp/argocd-local/ssh} ARGOCD_BINARY_NAME=argocd-application-controller $COMMAND --loglevel debug --redis localhost:${ARGOCD_E2E_REDIS_PORT:-6379} --repo-server localhost:${ARGOCD_E2E_REPOSERVER_PORT:-8081} --otlp-address=${ARGOCD_OTLP_ADDRESS} --application-namespaces=${ARGOCD_APPLICATION_NAMESPACES:-''} --server-side-diff-enabled=${ARGOCD_APPLICATION_CONTROLLER_SERVER_SIDE_DIFF:-'false'}"
```
Comment (rumstead) [inline Procfile:1]: Not sure if we need to make the same changes to the test tool `Procfile`?

**Comment by rumstead** [comment]
Didn't seem to help. When I run:
```
make install-test-tools-local
make start-e2e-local
make test-e2e-local
```
I see the `test-results` directory populated
```
❯ ls  test-results/cov* 
test-results/covcounters.59ea707707b901d5957baf3fc0fa64f0.12281.1715805185934105000 test-results/covcounters.c6bb65ce0a5fa66569a4cb1fc6efc6b4.40502.1717427615343989000
test-results/covcounters.59ea707707b901d5957baf3fc0fa64f0.12294.1715805187074782000 test-results/covcounters.c6bb65ce0a5fa66569a4cb1fc6efc6b4.40504.1717427620834574000
test-results/covcounters.c6bb65ce0a5fa66569a4cb1fc6efc6b4.39184.1717427697662708000 test-results/covcounters.c6bb65ce0a5fa66569a4cb1fc6efc6b4.40506.1717427621139324000
test-results/covcounters.c6bb65ce0a5fa66569a4cb1fc6efc6b4.39186.1717427417033107000 test-results/coverage.out
test-results/covcounters.c6bb65ce0a5fa66569a4cb1fc6efc6b4.40499.1717427617518722000 test-results/covmeta.59ea707707b901d5957baf3fc0fa64f0
test-results/covcounters.c6bb65ce0a5fa66569a4cb1fc6efc6b4.40500.1717427613681107000 test-results/covmeta.c6bb65ce0a5fa66569a4cb1fc6efc6b4
```

**Comment by rumstead** [comment]
Ok after looking at the `e2e-server.log` I see the error for all the binaries. 
```
19:20:11               repo-server | [merror: coverage meta-data emit failed: output directory "test-results" inaccessible (err: stat test-results: no such file or directory); no coverage data written
```

**Pull request closed without merging**